ttnn/ttnn/distributed/ttrun.py

```diff
@@ -0,0 +1,283 @@
+
+        sorted_ranks = sorted(ranks)
+        if sorted_ranks != list(range(len(ranks))):
+            raise ValueError(f"Ranks must be contiguous from 0. Got: {sorted_ranks}")
+
+        return bindings
+
+    @field_validator("mesh_graph_desc_path")
+    def validate_mesh_graph_exists(cls, path: str) -> str:
+        """Ensure mesh graph descriptor file exists"""
+        mesh_path = Path(path).expanduser().resolve()
+        if not mesh_path.exists():
+            raise ValueError(f"Mesh graph descriptor not found: {mesh_path}")
+        return str(mesh_path)
+
+
+def parse_binding_config(yaml_path: Path) -> TTRunConfig:
+    """Parse YAML configuration file with schema validation."""
+    if not yaml_path.exists():
+        raise ValueError(f"Configuration file not found: {yaml_path}")
+
+    with open(yaml_path, "r") as f:
+        data = yaml.safe_load(f)
+
+    try:
+        return TTRunConfig(**data)
+    except ValidationError as e:
+        raise ValueError(f"Invalid configuration: {e}")
+
+
+def get_rank_environment(binding: RankBinding, config: TTRunConfig) -> Dict[str, str]:
+    """Get all environment variables for a specific rank.
+
+    Args:
+        binding: Rank binding configuration
+        config: Global configuration
+
+    Returns:
+        Dictionary of environment variables for this rank
+    """
+    env = {
+        "PYTHONPATH": os.environ.get("PYTHONPATH", os.getcwd()),
+        "TT_METAL_HOME": os.environ.get("TT_METAL_HOME", os.getcwd()),
+        "TT_METAL_CACHE": os.environ.get(
+            "TT_METAL_CACHE",
+            DEFAULT_CACHE_DIR_PATTERN.format(home=str(Path.home()), hostname=os.uname().nodename, rank=binding.rank),
+        ),
+        "TT_MESH_ID": str(binding.mesh_id),
+        "TT_HOST_RANK": str(binding.host_rank_id),
+        "TT_MESH_GRAPH_DESC_PATH": config.mesh_graph_desc_path,
+    }
+
+    # Apply environment variables with expansion and proper precedence
+    # Global environment variables first
+    env.update({k: os.path.expandvars(v) for k, v in config.global_env.items()})
+    # Rank-specific overrides last (higher precedence)
+    env.update({k: os.path.expandvars(v) for k, v in binding.env_overrides.items()})
+
+    return env
+
+
+def build_rank_environment_args(binding: RankBinding, config: TTRunConfig) -> List[str]:
+    """Build environment variable arguments for mpirun.
+
+    Args:
+        binding: Rank binding configuration
+        config: Global configuration
+
+    Returns:
+        List of ["-x", "KEY=value"] arguments for mpirun
+    """
+    env_args = []
+    env = get_rank_environment(binding, config)
+
+    for key, value in env.items():
+        env_args.extend(["-x", f"{key}={value}"])
+
+    return env_args
+
+
+def build_mpi_command(config: TTRunConfig, program: List[str], mpi_args: Optional[List[str]] = None) -> List[str]:
+    """Build OpenMPI command with per-rank environment variables."""
+    cmd = ["mpirun"]
+
+    if mpi_args:
```

Comment:
Great use of Python type declarations, but note that we don't have `mypy` or any other static type checkers like that.

If you want to be REALLY verbose, you could do stuff like

```suggestion
    if mpi_args:
        assert isinstance(mpi_args, list)
        assert all(map(lambda arg_: isinstance(arg_, str), mpi_args)), f"mpi_args is not a list of strs"
```

** The code above may stink idk just an idea

Reply:
In this case, I'd prefer to push the problem to `mpirun` to deal with the problem if `mpi-args` is not well formed.